pkg/docker/config/config.go

```diff
@@ -0,0 +1,270 @@
+	registry = dockerHostNameToRegistry(registry)
+	return modifyJSON(func(auths *dockerConfigFile) (bool, error) {
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, setAuthToCredHelper(ch, registry, username, password)
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+		return true, nil
+	})
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+// If an entry is not found empty strings are returned for the username and password
+func GetAuthentication(registry string) (string, string, error) {
+	paths := [3]string{getPathToAuth(), filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName), filepath.Join(homedir.Get(), dockerLegacyCfg)}
+
+	for _, path := range paths {
+		username, password, err := findAuthentication(registry, path)
+		if err != nil {
+			return "", "", err
+		}
+		if username != "" && password != "" {
+			return username, password, nil
+		}
+	}
+	return "", "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = dockerHostNameToRegistry(registry)
+	return modifyJSON(func(auths *dockerConfigFile) (bool, error) {
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, deleteAuthFromCredHelper(ch, registry)
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return false, ErrNotLoggedIn
+		}
+		return true, nil
+	})
+}
+
+// dockerHostNameToRegistry converts the docker hostname to the docker registry
+func dockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() string {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName)
+	}
+	return filepath.Join(homedir.Get(), authCfg, authCfgFileName)
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file and returns it
+// or returns an empty dockerConfigFile data structure if auth.json does not exist
+// if the file exists and is empty, readJSONFile returns an error
+func readJSONFile(path string) (dockerConfigFile, error) {
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if path == filepath.Join(homedir.Get(), dockerLegacyCfg) {
```

Comment:
This would be better as a boolean parameter, and `GetAuthentication` with an `[]struct{path string, legacyFormat bool}`; that way we can change the path determination (e.g. a different `homedir` package) in a single place.

Reply:
fixed